reps/2022-04-21-state-observability-apis.md

```diff
@@ -0,0 +1,138 @@
+- Need to expose all necessary Ray states for basic observability.
+- State APIs should be consistent / stable / well-documented. E.g., APIs names must be consistent to the concepts
+- APIs should be available to all clients, such as Dashboard, CLI, and Python APIs with the same implementation.
+- Shouldn’t introduce high overhead to the normal workloads.
+- Under no circumstances should it be possible to crash a cluster (say, up to our scalability envelope) using the observability APIs.
+
+
+### Should this change be within `ray` or outside?
+
+Accessing state is a fundamental feature to debug systems. These changes would lie within Ray to improve the basic observability of the project.
+
+## Stewardship
+### Required Reviewers
+
+@scv119, @edoakes, @ericl
+
+### Shepherd of the Proposal (should be a senior committer)
+
+@edoakes
+
+## Design and Architecture
+
+### States
+“States” must be closely related and consistent to “concepts / APIs” users learn. For example, if we expose tasks as core Ray concepts, users would like to see the state of tasks. Ideally, the name of the APIs should be consistent with the concepts too.
+
+State APIs can be categorized by 3 types.
+
+**Logical states**
+- Aligned with logical concepts in which users need to learn to use Ray.
+- Useful from beginners to advanced users.
+- E.g., Actor, Task, Objects, Resources, Placement Group, Runtime Env
+
+**Physical states**
+- States from physical components.
+- When users scale up, it will become important to observe physical states. 
+- Logs, Cluster Health, Worker processes, Nodes
+
+**Internal states**
+- Internal information of Ray.
+- They are mostly useful for maintainers to debug bugs, but users can also be benefited. 
+- Scheduling states, Events, Internal states of Ray components 
+
+
+### Status quo
+The below table shows if the existing APIs support the states above. Each column explains if the API is available for each client (Python API, CLI, REST API). Each row indicates available states in Ray. 
+
+TODO(sang): Add a picture.
+
+### Proposed APIs
+All states should be visible to the users through properly documented / stable / consistently named APIs. This section proposes the state APIs to observe the following resources. 
+
+- Green: Already available. Might need to make APIs public/stable. 
+- Red, P0: States that cannot be accessed / hard to access now
+- Yellow, P0: API exists, but we need refinement or renaming.
+- Blue, P1: Less critical states. Could be missing or not refined.
+
+TODO(sang): Add a picture
+
+### Terminologies
+- Consumer: Consumers of state APIs. In Ray, (e.g., dashboard, CLI, and Python APIs).
+- API server: Currently a Ray dashboard. Centralized server to aggregate and provide state information of the cluster.
+- Ray agent: Currently a dashboard agent. A Python process running per node.
+- Resources: Objects created by Ray. E.g., tasks, actors, placement groups, namespace. 
+
+### Observability API architecture
```

Comment:
I think we should add a short summary of alternatives considered here (e.g., clients talking directly to each node).